src/clojure/nrepl/socket.clj

```diff
@@ -0,0 +1,239 @@
+
+(def orig-warn-on-reflection *warn-on-reflection*)
+
+(defmacro find-class [full-path]
+  `(try
+     (Class/forName (name ~full-path))
+     (catch ClassNotFoundException ex#
+       nil)))
+
+;;; InetSockets (TCP)
+
+(defn inet-socket [bind port]
+  (let [port (or port 0)
+        addr (fn [^String bind port] (InetSocketAddress. bind (int port)))
+        ;; We fallback to 127.0.0.1 instead of to localhost to avoid
+        ;; a dependency on the order of ipv4 and ipv6 records for
+        ;; localhost in /etc/hosts
+        bind (or bind "127.0.0.1")]
+    (doto (ServerSocket.)
+      (.setReuseAddress true)
+      (.bind (addr bind port)))))
+
+;; Unix domain sockets
+
+(def ^Class junixsocket-address-class
+  (find-class 'org.newsclub.net.unix.AFUNIXSocketAddress))
+
+(def ^Class junixsocket-server-class
+  (find-class 'org.newsclub.net.unix.AFUNIXServerSocket))
+
+(def ^Class jdk-unix-address-class
+  (find-class 'java.net.UnixDomainSocketAddress))
+
+(def ^Class jdk-unix-server-class
+  (find-class 'java.nio.channels.ServerSocketChannel))
+
+(def ^:private test-junixsocket?
+  ;; Make it possible to test junixsocket even when JDK >= 16
+  (= "true" (System/getProperty "nrepl.test.junixsocket")))
+
+(def unix-domain-flavor
+  (cond
+    test-junixsocket? (do
+                        (assert junixsocket-address-class)
+                        (assert junixsocket-server-class)
+                        (binding [*out* *err*]
+                          (println "nrepl.test: insisting on junixsocket support"))
+                        :junixsocket)
+    (and jdk-unix-address-class jdk-unix-server-class) :jdk
+    (and junixsocket-address-class junixsocket-server-class) :junixsocket
+    :else nil))
+
+(def jdk-unix-address-of
+  (when (= :jdk unix-domain-flavor)
+    (let [addr-of (.getDeclaredMethod jdk-unix-address-class "of"
+                                      (into-array Class [String]))]
+      (fn [path] (.invoke addr-of nil (into-array String [path]))))))
+
+(def junix-address-of
+  (when (= :junixsocket unix-domain-flavor)
+    (let [c (.getDeclaredConstructor junixsocket-address-class
+                                     (into-array Class [File]))]
+      (fn [path] (.newInstance c (into-array File [(File. ^String path)]))))))
+
+(defn unix-socket-address
+  "Returns a filesystem socket address for the given path string."
+  [^String path]
+  (case unix-domain-flavor
+    :jdk (jdk-unix-address-of path)
+    :junixsocket (junix-address-of path)
+    (let [msg "Support for filesystem sockets requires JDK 16+ or a junixsocket dependency"]
+      (log msg)
+      (throw (ex-info msg {:nrepl/kind ::no-filesystem-sockets})))))
+
+(set! *warn-on-reflection* false)
+
+;; SocketAddress doesn't have .getPath until JDK 16, and we can't refer to
+;; AFUnixSocketAddress unconditionally in the junixsocket cases.  Also note that
+;; the former returns a Path, and the latter returns a string.
+
+(defn- get-path [addr] (.getPath addr))
```

Comment:
I think something like this should work and address the nuances you mention, WDYT?

```clj
(defmacro getPath [addr]
  (if (Class/forName ...)
    `(let [^Path v# (.getPath (with-meta ~addr {:tag `com.Foo}))]
       ...)
    `(let [^String v# (.getPath (with-meta ~addr {:tag `com.Bar}))]
       ...)))
```

An important point is that all usages of the macro should perceive the same return type regardless of the `if` branch. e.g. always return string, or file, whatever, by using logic _within_ the macro.

This way, consumers of `get-path` can be abstracted away. Wouldn't look good to have to repeat assumptions/inspections in every usage of `get-path`; that would seem error-prone.

Reply:
Interesting, but I wonder if this will work with current or future AOT, i.e. for this to be correct, the macro would have to always re-expand at run time, not compile time, since the available classes (and decision about the socket flavor) aren't known until then.

Reply:
Very good observation :)

However shipping AOT-compiled libs tends to be very much frowned upon. I guess there can be exceptions, is nrepl one? @bbatsov

Reply:
I suppose, unless we think of something better (or unless the above concern isn't relevant), I'd be tempted to just leave the reflection override for now, or to remove it and live with the warning, whichever's preferable.

Reply:
Personally I'd be inclined to follow (or at least try) my suggestion, it should work. If you don't think it would work I'd appreciate a specific reason / your attempt

As for AOT:

* in principle I guess nrepl is not AOT-distributed, and/as most clojure libs never will or should
* there's the edge case of using nrepl in an AOT-compiled production app (for having a production repl)
  * however, for an issue to arise, the uberjar-er environment and the production environment should be executing different JDKs and/or computed classpaths
    *  this drift between "source" and "target" environments seems a notably bad practice, so having an error would nudge people into the right direction.

Reply:
I'm happy to handle it however's preferred, though personally, I'd lean against the extra complexity, since it's a rarely called function.  In the end, I may also need to defer to you with respect to whether or not it's safe enough in all relevant situations.

My assumption was that if you AOT compile the file, the macro will expand based on the compile time values, hard-coding the type coercions, which might or might not be correct at runtime.  Note that the "right" value can also be changed later, on any jvm invocation via the nrepl.test.junixsocket property.  Granted that shouldn't be used outside of testing, but it contributes to the additional complexity we have to think about, that we otherwise wouldn't if we didn't worry about the one reflection.

But, as mentioned, I'm happy to switch to the macro if you all feel confident it's sufficiently safe, and would prefer that.

Reply:
Yes I understood what you're saying, thanks! What I said about AOT remains applying

The grander context is that we've been eliminating reflection from a variety of projects, from nrepl, to the most prominent projects in https://github.com/clojure-emacs/, to the clojurescript compiler (WIP), to having a dedicated linter in Eastwood, etc.

It takes one warning to foil the goal of a reflection-free tooling environment that won't pollute end-users' repls when they opt for a global `*warn-on-reflection* true` setting.

In that way a "no broken windows" approach might be considered pragmatic: it would take one broken window to set the precedent that introducing warnings is OK after all the fixing effort.

As always, this is just my input (only partly qualified as I'm a new clojure-emacs member) so we can always hear out @bbatsov when in doubt 😁

Reply:
Another option I just thought of might be to mark the return as `^String`, which I think it should be, and handle any internal reflection manually, i.e. just rewrite get-path to look up the right method, etc.  Not beautiful, but it might work, and would avoid the need for the `set!` bits, etc.  I can try that if it seems interesting/desirable.

Reply:
Not entirely sure of what you mean. Feel free to try a few things though, it's good to have options 👍

If you mean doing reflection by hand (using Java's reflection API instead of Clojure's), while nominally it would "prevent" warnings, in practice it would be the same, if not slightly worse because these can't be automatedly detected.

I've personally seen usages of Java's reflection API creating actual issues, especially with newer JDKs which are sensitive around `reflective access`. Typically they print a warning at JDK level, but it is foreseen that they will just fail in newer JDKs.

Reply:
Nevermind - that won't work, I think, because we still don't know the actual class at compile time, and so even a reflection based `(.invoke method addr)` is going to cause a complaint because we don't (and can't) know the class of addr at compile time.

Reply:
Sorry for not joining the party here earlier - during my vacation I try to pay less attention to my projects (and it seems I succeed :D). 

> However shipping AOT-compiled libs tends to be very much frowned upon. I guess there can be exceptions, is nrepl one?

Currently it's not and I don't have any plans to ship an AOT version in the foreseeable future. I guess we just have document the impact of this solution to a potential AOT version of nREPL. At any rate - I'm not very concerned about this. Still, it would have been nice if this could have been solved in some simpler way, but I'm short on ideas.

Reply:
Adding a note in the documentation would seem a nice middle ground - definitely better than an hypothetical "you're using AOT, go figure it out" 😄